tests/unit/obuf-limits.tcl

```diff
@@ -70,4 +70,32 @@ start_server {tags {"obuf-limits"}} {
         assert {$omem >= 100000 && $time_elapsed < 6}
         $rd1 close
     }
+
+    test {Won't write replies if client output buffer hard limit is enforced} {
+        r config set client-output-buffer-limit {normal 100000 0 0}
+        # Total size of all items must be more than 100k
+        set item [string repeat "x" 1000]
+        for {set i 0} {$i < 150} {incr i} {
+            r lpush mylist $item
+        }
+        set orig_mem [s used_memory]
+        # Set client name and get all items
+        set rd [redis_deferring_client]
+        $rd client setname mybiglist
+        assert {[$rd read] eq "OK"}
+        $rd lrange mylist 0 -1
+        $rd flush
+        after 100
+
+        # Before we read reply, redis will close this client.
+        set clients [r client list]
+        assert_no_match "*name=mybiglist*" $clients
+        set cur_mem [s used_memory]
+        # 10k just is a deviation threshold
+        assert {$cur_mem < 10000 + $orig_mem}
```

Comment:
It never fails regardless of 100k or 10M of output buffer limit on both fixed and unfixed version, because we actually close `async-free` clients in`beforeSleep`, so no buffer will be kept in redis even if we don't read the fd. I initially thought you want me to test it is right on my commit.

Reply:
I'm confused. i thought that what we're fixing is that redis kept the partial response (what it cached up to the output buffer limit) in memory and writes it to the socket.
but now i realize it didn't keep it for very long.

in fact, for the purpose of our test, it was enough to just change the order of these lines in beforeSleep:
```c
    /* Handle writes with pending output buffers. */
    handleClientsWithPendingWritesUsingThreads();

    /* Close clients that need to be closed asynchronous */
    freeClientsInAsyncFreeQueue();
```

still your fix might be slightly better, but in order to write an assert that will show that in `used_memory`, we need the two commands to be handled in the same event loop, and in the right order. (we can probablu do that, but it's very complicated)

Reply:
There are two point i fixed

1. Don't continue to add reply bulk to output buffer(even if the reply is incomplete) to avoid increasing memory when reached output buffer limit.
2. Don't write any reply(regardless of complete or incomplete) to client to avoid unnecessary `write` cost.

For point 2, Changing the order of above two lines also won't write replies to client only when disable multi-threaded I/O, but this way also have a side effect the client flagged `CLIENT_CLOSE_AFTER_REPLY` will be closed with delay.

> we need the two commands to be handled in the same event loop, and in the right order. (we can probablu do that, but it's very complicated)

Exactly yes, we have talked about relate to this problem. I can't find a way to stably generate this case. I did a test without tcl test, I push several GB items to a big list, and set a small output buffer limit, i didn't found redis used much memory by `top` command when i executed `lrange mylist 0 -1`. Generating a several GB reply gave me enough time to observe this all, but i don't think we should do that in tcl tests.
For this assertion, maybe i didn't understand you, i initially thought you ask me to test my commit can release the output buffer memory without introducing a new bug.

Reply:
> For this assertion, maybe i didn't understand you, i initially thought you ask me to test my commit can release the output buffer memory without introducing a new bug.

i thought this assertion was there to make sure the client output buffer was released before the `read` executed. for some reason i thought that in the old code it wasn't the case.